src/v/cloud_storage/recursive_directory_walker.cc

```diff
@@ -49,18 +50,20 @@ ss::future<walk_result> recursive_directory_walker::walk(
 
         try {
             ss::file target_dir = co_await open_directory(target);
+            size_t files_in_dir{0};
             auto sub = target_dir.list_directory(
               [&files,
                &current_cache_size,
                &dirlist,
+               &files_in_dir,
                _target{target},
                _tracker{tracker}](ss::directory_entry entry) -> ss::future<> {
                   auto target{_target};
                   auto tracker{_tracker};
-                  vlog(cst_log.debug, "Looking at directory {}", target);
 
                   auto entry_path = std::filesystem::path(target)
                                     / std::filesystem::path(entry.name);
+                  files_in_dir++;
```

Comment:
it looks like this code was written before we were all fully enlightened on the dangers of lambda coroutines. sorta seems like either we should also bring the references down onto the stack, or `co_await` on `sub` directly and use `ss::coroutine::lambda` wrapper?

Reply:
I was also a little curious about why this function seems to work while committing the sin of capturing references into a coroutine lambda.

I think it's because this lambda gets moved into a stream as its `next` handler, and thereby kept alive as long as it will be called: this is different to a usual continuation use case where the lambda gets called to generate a future, and then the original lambda is thrown away.

However, that means we're depending on the implementation a bit too much and I'd like to stick with our general pattern of "don't reference capture lambdas!", so I've added a ss::coroutine::capture here anyway as a defensive measure in case the implementation of list_directory changes or this function gets moved around.